### PR TITLE
[EasyCore] Add change sets to events dispatched by `DeferredEntityEventDispatcher`

### PR DESCRIPTION
**Pull request by mvlad**
We need to know what changes were made after an entity was updated. To be consistent we going to do the same for entity creations and deletions.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

**Comment by mvlad** [comment]
@roman-eonx @itorgov please see an alternative implementation - https://github.com/eonx-com/easy-monorepo/commit/b765f5b8185ac1529ad38592fd3c2614a7dbf89e

**Comment by roman-eonx** [comment]
> @roman-eonx @itorgov please see an alternative implementation - [b765f5b](https://github.com/eonx-com/easy-monorepo/commit/b765f5b8185ac1529ad38592fd3c2614a7dbf89e)

Looks good. But I'd keep the method names. The initial idea of this naming was: we "defer" first and "dispatch" at the end.

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php [hunk 1]
```diff
@@ -36,9 +41,28 @@ public function __construct(EventDispatcherInterface $eventDispatcher)
         $this->eventDispatcher = $eventDispatcher;
     }
 
+    /**
+     * @param int $transactionNestingLevel
```
Comment (itorgov) [inline packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:45]: You have defined these params in the interface. Do we really need to define them twice?

File: packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriber.php [hunk 2]
```diff
@@ -48,11 +57,19 @@ public function onFlush(OnFlushEventArgs $eventArgs): void
         $scheduledEntityUpdates = $this->filterEntities($unitOfWork->getScheduledEntityUpdates());
```
Comment (itorgov) [inline packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriber.php:57]: You have forgot to call `computeChangeSets`.

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php [hunk 3]
```diff
@@ -1,359 +0,0 @@
...
-use EonX\EasyCore\Doctrine\Events\EntityUpdatedEvent;
-use EonX\EasyCore\Tests\AbstractTestCase;
-use Prophecy\Argument;
-use Psr\EventDispatcher\EventDispatcherInterface;
-use stdClass;
-
-/**
- * @covers \EonX\EasyCore\Doctrine\Dispatchers\DeferredEntityEventDispatcher
- */
-final class DeferredEntityEventDispatcherTest extends AbstractTestCase
```
Comment (itorgov) [inline packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:18]: Why did you remove it?

**Review by itorgov — CHANGES_REQUESTED** [review]

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php (hunk 1 above)
Comment (mvlad) [reply packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php]: These params have been removed along with the method.

File: packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriber.php (hunk 2 above)
Comment (mvlad) [reply packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriber.php:57]: Why do we need the call here?
As I see, it's already called in `\Doctrine\ORM\UnitOfWork::commit`:
1. Dispatch `preFlush`
2. `computeChangeSets`
3. Dispatch `onFlush` (out logic is here)
4. Begin and commit transaction

File: packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php (hunk 3 above)
Comment (mvlad) [reply packages/EasyCore/tests/Doctrine/Dispatchers/DeferredEntityEventDispatcherTest.php:18]: I did it for the following reasons:
1. Now `EntityEventSubscriberTest` is the integration test that covers all cases that were in `DeferredEntityEventDispatcherTest`.
2. It's difficult to maintain `DeferredEntityEventDispatcherTest` because there are a lot of implementation details.

But if you think that `DeferredEntityEventDispatcherTest` can catch some errors, I will return it.

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php [hunk 4]
```diff
@@ -39,90 +44,123 @@ public function __construct(EventDispatcherInterface $eventDispatcher)
...
+        if (isset($this->entityInsertions[$oid]) !== false) {
+            return new EntityCreatedEvent($this->entityInsertions[$oid], $entityChangeSet);
+        }
+
+        if (isset($this->entityUpdates[$oid]) !== false) {
+            return new EntityUpdatedEvent($this->entityUpdates[$oid], $entityChangeSet);
+        }
+
+        // @codeCoverageIgnoreStart
+        throw new \LogicException('Entity fot event not found');
```
Comment (roman-eonx) [inline packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:144]: Please fix the typo and add a point at the end of the message.

**Review by roman-eonx — CHANGES_REQUESTED** [review]

File: packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriber.php (hunk 2 above)
Comment (itorgov) [reply packages/EasyCore/src/Doctrine/Subscribers/EntityEventSubscriber.php:57]: Ok

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php [hunk 5]
```diff
@@ -39,90 +44,123 @@ public function __construct(EventDispatcherInterface $eventDispatcher)
...
-        $this->entityUpdates[$transactionNestingLevel] = $mergedEntityUpdates;
+        $oid = \spl_object_hash($object);
+        $this->entityUpdates[$oid] = $object;
+        $this->entityChangeSets[$transactionNestingLevel][$oid] = $entityChangeSet;
     }
 
     public function disable(): void
     {
         $this->enabled = false;
+        $this->clear();
```
Comment (itorgov) [inline packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:92]: We don't have to clear it during disabling. Just imagine the following flow:

- Start a transaction.
- Create the entity A.
- Disable the dispatcher.
- Update the entity B.
- Enable the dispatcher.
- Commit the transaction.

This way we should dispatch an event about the entity A.

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php [hunk 6]
```diff
@@ -39,90 +44,123 @@ public function __construct(EventDispatcherInterface $eventDispatcher)
...
-                }
-                $processedEntities[$oid] = $entity;
-                $this->eventDispatcher->dispatch(new EntityUpdatedEvent($entity));
+            foreach ($mergedEntityChangeSets as $oid => $entityChangeSet) {
+                $event = $this->createEntityEvent($oid, $entityChangeSet);
+
+                $this->eventDispatcher->dispatch($event);
             }
+        } finally {
+            $this->clear();
```
Comment (itorgov) [inline packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:118]: Why does this approach better then the previous one?

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php [hunk 7]
```diff
@@ -39,90 +44,123 @@ public function __construct(EventDispatcherInterface $eventDispatcher)
     public function clear(?int $transactionNestingLevel = null): void
     {
         if ($transactionNestingLevel !== null) {
```
Comment (itorgov) [inline packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:46]: Why don't you clear the `$this->entityInsertions` and `$this->entityUpdates` arrays for the specified transaction level?

**Review by itorgov — CHANGES_REQUESTED** [review]

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php (hunk 6 above)
Comment (mvlad) [reply packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:118]: 1. I need `entityInsertions` and `entityUpdates` in a separate method `createEntityEvent`. In the previous approach, these fields are empty.
2. This approach is a bit more optimal as it avoids a temporary copy of fields.

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php (hunk 7 above)
Comment (mvlad) [reply packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:46]: 1. Now these fields don't contain transaction levels. These fields are needed only to choose the event type and construct it.
2. It's able to add transaction levels to these fields. But this will complicate the code, and I see no benefit for this (please correct me if I'm wrong).

Comment (itorgov) [reply packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:46]: OK, but this way we need to add:
```
$this->entityInsertions = [];
$this->entityUpdates = [];
```
right before the early return in the `if` block.

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php (hunk 4 above)
Comment (mvlad) [reply packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php]: Fixed.

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php (hunk 5 above)
Comment (mvlad) [reply packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php]: Fixed.

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php (hunk 7 above)
Comment (mvlad) [reply packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:46]: We can' add these lines before the early return in the `if` because [testEventsAreDispatchedWhenExceptionIsThrownAndCatched](https://github.com/eonx-com/easy-monorepo/blob/490ba6543df82f422bce9985659864290a68d625/packages/EasyCore/tests/Doctrine/Subscribers/EntityEventSubscriberTest.php#L68) will not pass.
Wrote a logic to clean up `entityInsertions`/`entityUpdates`.

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php [hunk 8]
```diff
@@ -93,36 +112,73 @@ public function disable(): void
...
+        if (isset($this->entityInsertions[$oid]) !== false) {
+            return new EntityCreatedEvent($this->entityInsertions[$oid], $entityChangeSet);
+        }
+
+        if (isset($this->entityUpdates[$oid]) !== false) {
+            return new EntityUpdatedEvent($this->entityUpdates[$oid], $entityChangeSet);
+        }
+
+        // @codeCoverageIgnoreStart
+        throw new \LogicException('Entity for event not found.');
```
Comment (roman-eonx) [inline packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:162]: This class needs to be imported.

**Review by roman-eonx — CHANGES_REQUESTED** [review]
Let's rework this PR by introducing a new `EasyDoctrine` package and copying the contents of `easy-core/src/Doctrine` to it:
- Move `FixPostgreSqlDefaultSchemaListener` from `Events` to `Listeners`.
- Don't copy `ConnectionChecker`.
- Deprecate the original classes (similar to how it's done [here](https://github.com/eonx-com/easy-core/blob/master/src/Doctrine/DBAL/ConnectionChecker.php#L10)).
- Apply the changes from this PR only to the classes of the new package.
- Update the PR description accordingly.

File: packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php (hunk 8 above)
Comment (mvlad) [reply packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php:162]: Fixed in `EasyDoctrine` package - https://github.com/eonx-com/easy-monorepo/blob/33c75a99d89874fa21b7e1e78c8fca85c6a37332/packages/EasyDoctrine/src/Dispatchers/DeferredEntityEventDispatcher.php#L9

**Comment by mvlad** [comment]
New PR - https://github.com/eonx-com/easy-monorepo/pull/775
